Content.Server/UserInterface/ActivatableUISystem.cs

```diff
@@ -0,0 +1,143 @@
+            SubscribeLocalEvent<ActivatableUIComponent, HandDeselectedEvent>((uid, aui, _) => CloseAll(uid, aui));
+            SubscribeLocalEvent<ActivatableUIComponent, UnequippedHandEvent>((uid, aui, _) => CloseAll(uid, aui));
+        }
+
+        private void OnActivate(EntityUid uid, ActivatableUIComponent component, ActivateInWorldEvent args)
+        {
+            if (args.Handled) return;
+            if (component.InHandsOnly) return;
+            args.Handled = InteractInstrument(args.User, component);
+        }
+
+        private void OnUseInHand(EntityUid uid, ActivatableUIComponent component, UseInHandEvent args)
+        {
+            if (args.Handled) return;
+            args.Handled = InteractInstrument(args.User, component);
+        }
+
+        private bool InteractInstrument(IEntity user, ActivatableUIComponent aui)
+        {
+            if (!user.TryGetComponent(out ActorComponent? actor)) return false;
+
+            if (aui.AdminOnly && !_adminManager.IsAdmin(actor.PlayerSession)) return false;
+
+            if (!_actionBlockerSystem.CanInteract(user.Uid))
+            {
+                user.PopupMessageCursor(Loc.GetString("base-computer-ui-component-cannot-interact"));
+                return true;
+            }
+
+            var ui = aui.UserInterface;
+            if (ui == null) return false;
+
+            if (aui.SingleUser && (aui.CurrentSingleUser != null) && (actor.PlayerSession != aui.CurrentSingleUser))
+            {
+                // If we get here, supposedly, the object is in use.
+                // Check with BUI that it's ACTUALLY in use just in case.
+                // Since this could brick the object if it goes wrong.
+                if (ui.SubscribedSessions.Count != 0) return false;
+            }
+
+            // If we've gotten this far, fire a cancellable event that indicates someone is about to activate this.
+            // This is so that stuff can require further conditions (like power).
+            var oae = new ActivatableUIOpenAttemptEvent(user);
+            RaiseLocalEvent(aui.OwnerUid, oae, false);
+            if (oae.Cancelled) return false;
+
+            SetCurrentSingleUser(aui.OwnerUid, actor.PlayerSession, aui);
+            ui.Toggle(actor.PlayerSession);
+            return true;
+        }
+
+        public void SetCurrentSingleUser(EntityUid uid, IPlayerSession? v, ActivatableUIComponent? aui = null)
+        {
+            if (!Resolve(uid, ref aui))
+                return;
+            if (!aui.SingleUser)
+                return;
+            if (aui.CurrentSingleUser != null)
+                aui.CurrentSingleUser.PlayerStatusChanged -= aui.OnPlayerStatusChanged;
+
+            aui.CurrentSingleUser = v;
+
+            if (v != null)
+                v.PlayerStatusChanged += aui.OnPlayerStatusChanged;
+            RaiseLocalEvent(uid, new ActivatableUIPlayerChangedEvent(), false);
+        }
```

Comment:
Again, **IF** it's possible to just use `BoundUIClosedEvent`, this function could probably be simplified quite a bit. 

Also I believe `ActivatableUIPlayerChangedEvent` would become unnecessary. The `clean` function could just be called by the instrument system subscribing to `BoundUIClosedEvent`

Reply:
ActivatableUIPlayerChangedEvent is specifically for *the* single user of the device, I'm still not sure if this is going to be expanded to allow "observation" of the device at a future time without being a user of it.

Reply:
Okay yeah, if you want spectators and need to distinguish between the "user" and some generic client with the UI open I don't see an easy way to do that without some sort of `ActivatableUIPlayerChangedEvent`